frontend/src/lib/k8s/api/v2/hooks.ts

```diff
@@ -0,0 +1,192 @@
+import { useQuery, useQueryClient } from '@tanstack/react-query';
+import { useEffect, useMemo, useRef } from 'react';
+import { getCluster } from '../../../cluster';
+import { ApiError, QueryParameters } from '../../apiProxy';
+import { KubeObjectClass, KubeObjectInterface } from '../../cluster';
+import { clusterFetch } from './fetch';
+import { KubeList } from './KubeList';
+import { KubeObjectEndpoint } from './KubeObjectEndpoint';
+import { makeUrl } from './utils';
+import { watchList, watchObject } from './watching';
+
+/**
+ * Returns a single KubeObject.
+ */
+export function useKubeObject<T extends KubeObjectClass>({
+  kubeObjectClass,
+  namespace,
+  name,
+  cluster: maybeCluster,
+}: {
+  kubeObjectClass: T;
+  namespace?: string;
+  name: string;
+  cluster?: string;
+}) {
+  type Instance = InstanceType<T>;
+  const endpoint = useEndpoints(kubeObjectClass.apiEndpoint.apiInfo);
+  const cluster = maybeCluster ?? getCluster() ?? '';
+
+  const queryKey = useMemo(
+    () => ['object', cluster, endpoint, namespace, name],
+    [endpoint, namespace, name]
+  );
+
+  const socketRef = useRef<Promise<WebSocket> | undefined>();
+  function startWatching() {
+    if (socketRef.current || !endpoint) return;
+
+    socketRef.current = watchObject({
+      endpoint,
+      name,
+      onObject: obj => client.setQueryData<KubeObjectInterface>(queryKey, obj),
+      itemClass: kubeObjectClass,
+    });
+  }
+
+  useEffect(() => {
+    if (!endpoint) return;
+    const maybeData = client.getQueryData<KubeList<Instance>>(queryKey);
+    if (maybeData) {
+      startWatching();
+    }
+
+    return () => {
+      socketRef.current?.then(socket => socket.close());
+    };
+  }, [endpoint]);
+
+  const client = useQueryClient();
+  const query = useQuery<Instance | null, ApiError>({
+    enabled: !!endpoint,
+    placeholderData: null,
+    staleTime: 5000,
+    queryKey,
+    queryFn: async () => {
+      if (!endpoint) return;
+      const url = makeUrl([KubeObjectEndpoint.toUrl(endpoint, namespace), name]);
+      const obj: KubeObjectInterface = await clusterFetch(url, {
+        cluster,
+      }).then(it => it.json());
+      startWatching();
+      return new kubeObjectClass(obj);
+    },
+  });
+
+  return { ...query, data: query.error ? null : query.data ?? null };
+}
+
+/**
+ * Test different endpoints to see which one is working.
+ */
+const getWorkingEndpoint = async (endpoints: KubeObjectEndpoint[]) => {
+  const promises = endpoints.map(endpoint => {
+    return clusterFetch(KubeObjectEndpoint.toUrl(endpoint), {
+      method: 'GET',
+      cluster: getCluster() ?? '',
+    }).then(it => {
+      if (!it.ok) {
+        throw new Error('error');
+      }
+      return endpoint;
+    });
+  });
+  return Promise.any(promises);
+};
+
+const useEndpoints = (endpoints: KubeObjectEndpoint[]) => {
```

Comment:
This one returns a useQuery type. Can you please consider abstracting it? Perhaps creating a new type name from the useQuery one if that's the only option.

(same for any other ones in this file I may have missed)

Reply:
I agree here. Unless we are limiting too much what can be done, I'd keep the exposure to react-query to a minimum so we don't have to compromise on an implementation detail.

Reply:
done